### DISCUSSION TITLE
Is it difficult to add multiprecision suport? Any guidelines?

### DISCUSSION THREAD

**Discussion by srogatch** [osqp/osqp#640] [Ideas]
I'm currently using OSQP solver in this repo https://github.com/srogatch/bsat-circulation/blob/lin-prog/CpuSolver.h to solve Boolean Satisfiability according to this article https://www.academia.edu/118325445/P_NP_a_reduction_of_Boolean_Satisfiability_problem_to_Linear_Programming_and_Long_Arithmetic , also discussed here: https://cstheory.stackexchange.com/questions/54250/can-you-find-a-counterexample-disprove-my-p-np-solution .

So to solve BSAT in polynomial time and test my algorithm (probably disprove or fix if it doesn't converge for thousands of variables), I need to add multiprecision floating- or fixed-point arithmetic to OSQP. Would you accept such a pull request? Any guidelines on how to do this quickly?

Important note: it's enough that just the objective function is optimized in multi-precision / long arithmetic. Is there a shortcut to convert just the objective function calculation to multi-precision, without converting every computation in OSQP to multi-precision too? This would save a lot of runtime, because for a million of variables in the boolean formula, the numbers would be 250KB each.
[BSAT-Polynomial-Objective.pdf](https://github.com/osqp/osqp/files/15168503/BSAT-Polynomial-Objective.pdf)
[RU-BSAT-Polynomial-Objective.pdf](https://github.com/osqp/osqp/files/15168512/RU-BSAT-Polynomial-Objective.pdf)

**Comment by imciner2** [comment]
What do you mean by multi precision in this case? Do you mean mixed-precision computations (i.e. using one precision in the linear system solver, and another in the vector updates), or do you mean replacing all numbers with quad/double-double arithmetic?

If it is a higher precision arithmetic (like quad/double-double), then that is actually easier because we can just redefine `OSQPFloat` to be a different type, and OSQP can build with that precision then - at least I think that should be the case for the built-in algebra and the QDLDL solver.  (note that it will require you to pass in all values in that precision as well).

If it is the former (mixed-precision), then that is a lot more involved.

> it's enough that just the objective function is optimized in multi-precision / long arithmetic. Is there a shortcut to convert just the objective function calculation to multi-precision, without converting every computation in OSQP to multi-precision too? 

At this point, I don't think we could split the cost function into separate precisions from the rest, because we merge the constraints with the cost to form the KKT, and then solve that, so we would need the same precision in the constraint matrices as the cost function.

**Reply by srogatch** [reply]
Thanks a lot, this is very helpful! Apparently, there's `__float128` data type that I can try to swap in. This would allow testing up to about 50 variables. And then there's Boost Multiprecision that should be easy to swap in like a native data type, right?

**Reply by srogatch** [reply]
Apparently, it's not that easy to change to `__float128` @imciner2 - after I fixed it in OSQP and QDLDL, I'm getting weird numbers - see the attached screenshot:
![Screenshot from 2024-04-30 23-18-36](https://github.com/osqp/osqp/assets/5251612/2d544033-92d6-4248-b5de-9e4d70bb5c47)

Specifically, I did this: https://github.com/srogatch/osqp/commits/master/
Plus, after configuring with `cmake` I had to replace `double` with `__float128` in `algebra/_common/lin_sys/qdldl/qdldl_sources/CMakeLists.txt`. Still, something is wrong.

**Reply by goulart-paul** [reply]
Even the settings in output header are corrupted there.   I don't think the issue is QDLDL specific.

I think getting this to work would be quite difficult.   Have you considered trying this in Julia via COSMO.jl?   It is nearly the same method as OSQP for QPs but much more flexibly typed.

**Reply by srogatch** [reply]
Yet another language for me to learn - thanks for the suggestion, I'll weigh in on the best way to proceed...

**Reply by imciner2** [reply]
Are you even getting weird numbers when reading them directly in your solver? I might expect the printed values by OSQP to be weird, because I don't know if the printf function we use will have the proper format string to work with non-single/double inputs (we give it the float string, but that might not be enough for it).

But yes, it might be best to explore the pure Julia-based solvers to see how they will work for this. They can usually handle higher-precision data types transparently if you just give them data in that higher precision type.

**Reply by srogatch** [reply]
On the first iteration, I get a solution from your solver, but on the second your solver never converges. And I'm getting weird output even after replacing something like `printf("%f", x)` with `printf("%lf", (double)x)` so that `__float128` is converted to `double` before printing. Probably something in the codegen needs fixing, and I see `%.20f` hard-coded - so not good for 128-bit precision.

**Reply by srogatch** [reply]
@imciner2 @goulart-paul can we convert this to an issue that substitution of a `__float128` or an arbitrary-precision datatype (like Boost multiprecision) doesn't work?

**Comment by srogatch** [comment]
Oh, so there's a Python interface for COSMO.jl - that makes things easier for me:
https://github.com/oxfordcontrol/COSMO.jl
https://github.com/oxfordcontrol/cosmo-python

**Reply by srogatch** [reply]
I finally made it work without adaptive RHO: https://github.com/srogatch/osqp
So the problem seems to be in the RHO adaptation. Can you check, please?

**Reply by srogatch** [reply]
Got it: we have to ensure that constants are `__float128` too.
